osgpkitools/utils.py

```diff
@@ -58,3 +133,133 @@ def safe_write(filename, contents):
     """
     safe_rename(filename)
     atomic_write(filename, contents)
+
+def check_permissions(path):
+    """The function checks for write permissions for the given path to verify if the user has write permissions
+    """
+    if os.access(path, os.W_OK):
+        return
+    else:
+        raise FileWriteException("User does not have appropriate permissions for writing to current directory.")
+
+
+def find_user_cred(usercert=None, userkey=None):
+    """Find a readable user cert/key pair, trying pairs in the following order:
+    1. usercert, userkey
+    2. X509_USER_CERT, X509_USER_KEY environment variables
+    3. '~/.globus/usercert.pem', '~/.globus/userkey.pem'
+    INPUT (optional)
+    usercert: path to user certificate
+    userkey: path to private key of user
+    OUTPUT
+    Paths to the user cert and key
+    """
+    
+    # list of cert/key pairs to try
+    input_pairs = [(usercert, userkey),
+                   ((os.environ.get('X509_USER_CERT'), os.environ.get('X509_USER_KEY'))),
+                   (os.path.expanduser('~/.globus/usercert.pem'), os.path.expanduser('~/.globus/userkey.pem'))]
+    cert_key_pairs = [t for t in input_pairs if None not in t] # remove undefined pairs for an improved err msg below
+
+    # M2Crypto doesn't raise exceptions when encountering missing or unreadable
+    # cert/key pairs so we force the issue
+    for cert, key in cert_key_pairs:
+        try:
+            open(cert, 'r')
+            open(key, 'r')
+            return cert, key
+        except IOError:
+            continue
+    raise IOError("Unable to read the following certificate/key pairs:\n- %s" %
+                  "\n- ".join([", ".join(pair) for pair in cert_key_pairs]))
+
+
+def print_failure_reason_exit(data):
+    """This functions prints the failure reasons and exits"""
+    try:
+        msg = 'The request has failed for the following reason: %s' % \
+        json.loads(data)['detail'].split('--')[1].lstrip()
+    except IndexError:
+        msg = 'The request has failed for the following reason: %s' % json.loads(data)['detail'].lstrip() + \
+              'Status : %s ' % json.loads(data)['status']
+
+    separator = '='*80
+    sys.exit('\n'.join(textwrap.wrap(separator + msg, width=80)))
+
+
+class Cert(object):
```

Comment:
I'd prefer that this ill-named class just go into request.py and integrate with the `write_csr` function already there

Reply:
But that would add a new dependency in that incommon_request.py would have to import from request.py.  That's not good, is it?  request.py is the osg-cert-request main, and that tool doesn't need to use this class.

Reply:
I think it's fine if the class goes into `request.py` and `incommon_request.py` has `from request import Cert`. It just makes a lot more sense for the CSR bits to go into the request library than utils.

But at the very least, let's rename the class to `CertSigningRequest` or `Csr` to better reflect what it handles/does.

Reply:
I will rename to Csr for now.  I agree this code should be shared between the two tools.  Please make an issue for @jteheran to consolidate that code later.  I don't like it being in request though, since that includes other things to support osg-cert-request.  Maybe it needs a new source file, like cert-utils.py or something like that.

Reply:
I'd be happy to move the `osg-cert-request` specific stuff out of `request.py` and have `request.py` just contain common cert request stuff.

Reply:
Ok, but can we do that later,  make an issue about it?